warehouse/attestations/services.py

```diff
@@ -0,0 +1,227 @@
+        used to verify the attestations.
+        Only GitHub Actions Trusted Publishers are supported.
+        """
+        publisher: OIDCPublisher | None = request.oidc_publisher
+        if not publisher or not publisher.publisher_name == "GitHub":
+            raise AttestationUploadError(
+                "Attestations are only supported when using Trusted "
+                "Publishing with GitHub Actions.",
+            )
+
+        try:
+            attestations = TypeAdapter(list[Attestation]).validate_json(
+                request.POST["attestations"]
+            )
+        except ValidationError as e:
+            # Log invalid (malformed) attestation upload
+            self.metrics.increment("warehouse.upload.attestations.malformed")
+            raise AttestationUploadError(
+                f"Error while decoding the included attestation: {e}",
+            )
+
+        if len(attestations) > 1:
+            self.metrics.increment(
+                "warehouse.upload.attestations.failed_multiple_attestations"
+            )
+
+            raise AttestationUploadError(
+                "Only a single attestation per file is supported.",
+            )
+
+        verification_policy = publisher.publisher_verification_policy(
+            request.oidc_claims
+        )
+        for attestation_model in attestations:
+            try:
+                predicate_type, _ = attestation_model.verify(
+                    Verifier.production(),
+                    verification_policy,
+                    distribution,
+                )
+            except VerificationError as e:
+                # Log invalid (failed verification) attestation upload
+                self.metrics.increment("warehouse.upload.attestations.failed_verify")
+                raise AttestationUploadError(
+                    f"Could not verify the uploaded artifact using the included "
+                    f"attestation: {e}",
+                )
+            except Exception as e:
+                with sentry_sdk.push_scope() as scope:
+                    scope.fingerprint = [e]
+                    sentry_sdk.capture_message(
+                        f"Unexpected error while verifying attestation: {e}"
+                    )
+
+                raise AttestationUploadError(
+                    f"Unknown error while trying to verify included attestations: {e}",
+                )
+
+            if predicate_type != AttestationType.PYPI_PUBLISH_V1:
+                self.metrics.increment(
+                    "warehouse.upload.attestations.failed_unsupported_predicate_type"
+                )
+                raise AttestationUploadError(
+                    f"Attestation with unsupported predicate type: {predicate_type}",
+                )
+
+        return attestations
+
+    def generate_and_store_provenance_file(
+        self, oidc_publisher: OIDCPublisher, attestations: list[Attestation], file: File
+    ) -> None:
```

Comment:
I'd like to separate concerns here a little bit and split this into two functions on the service, `generate_provenance` and `persist_provenance`.